File: problems/python/17. Letter Combinations of a Phone Number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []

        dmap = {
            "2":"abc", 
            "3":"def", 
            "4":"ghi", 
            "5":"jkl", 
            "6":"mno", 
            "7":"pqrs", 
            "8":"tuv", 
            "9":"wxyz"}
        
        q = [""]

        for d in digits:
            n = len(q)
            for _ in range(n):
                path = q.pop(0)
                ltrs = dmap[d]
                for c in ltrs:
                    q.append(path + c)
        
        return q

File: problems/Rust/17. Letter Combinations of a Phone Number.rs
use std::collections::HashMap;

impl Solution {
    pub fn letter_combinations(digits: String) -> Vec<String> {
        if digits.is_empty() {
            return vec![];
        }

        let dmap: HashMap<char, &str> = [
            ('2', "abc"), 
            ('3', "def"), 
            ('4', "ghi"), 
            ('5', "jkl"),
            ('6', "mno"), 
            ('7', "pqrs"), 
            ('8', "tuv"), 
            ('9', "wxyz"),
        ].iter().cloned().collect();
        
        let mut res = vec![String::new()];

        for d in digits.chars() {
            let mut temp = Vec::new();
            let ltrs = dmap[&d];
            for path in &res {
                for c in ltrs.chars() {
                    temp.push(format!("{}{}", path, c));
                }
            }
            res = temp;
        }
        res
    }
}